### PR TITLE
Add GetApplicableTrustPolicy function

### PR DESCRIPTION
**Pull request by rgnote**
Followed the rules from https://github.com/notaryproject/notaryproject/blob/main/trust-store-trust-policy-specification.md#selecting-a-trust-policy-based-on-artifact-uri

Signed-off-by: rgnote <5878554+rgnote@users.noreply.github.com>

### CODE REVIEW

**Review by gokarnm — CHANGES_REQUESTED** [review]
General comment, we should use terminology matching the spec, otherwise it gets confusing sometimes.

* trustPolicy instead of policyStatement
* (signature) verificationLevel vs  verificationPreset

File: verification/policy.go [hunk 1]
```diff
@@ -203,3 +203,27 @@ func ValidatePolicyDocument(policyDoc *PolicyDocument) error {
...
 }
+
+// GetApplicableTrustPolicy returns a pointer to the TrustPolicy statement that applies to the given
+// registry scope. If no applicable trust policy is found, returns an error
+func GetApplicableTrustPolicy(registryScope string, policyDoc *PolicyDocument) (*TrustPolicy, error) {
+
+	var wildcardPolicy *TrustPolicy
+	var applicablePolicy *TrustPolicy
+	for _, policyStatement := range policyDoc.TrustPolicies {
+		localCopy := policyStatement
```
[the remark this answers is not in the document]
Comment (rgnote) [reply verification/policy.go:214]: This line is deep copying. Deep copy will copy slice fields as well. https://go.dev/play/p/E3pDxriD5pn

File: verification/policy.go [hunk 2]
```diff
@@ -203,3 +203,27 @@ func ValidatePolicyDocument(policyDoc *PolicyDocument) error {
...
+			applicablePolicy = &localCopy
+		}
+	}
+
+	if applicablePolicy != nil {
+		return applicablePolicy, nil
+	} else if wildcardPolicy != nil {
+		return wildcardPolicy, nil
+	} else {
+		return nil, fmt.Errorf("registry scope %q has no applicable trust policy", registryScope)
```
[the remark this answers is not in the document]
Comment (rgnote) [reply verification/policy.go:227]: why not an error?

File: verification/policy.go (hunk 1 above)
[the remark this answers is not in the document]
Comment (rgnote) [reply verification/policy.go]: Discussed offline. Replaced it with real deep copy functionality.

File: verification/policy.go [hunk 3]
```diff
@@ -203,3 +203,48 @@ func ValidatePolicyDocument(policyDoc *PolicyDocument) error {
...
+func (policyDoc *PolicyDocument) getApplicableTrustPolicy(registryUri string) (*TrustPolicy, error) {
+	i := strings.LastIndex(registryUri, ":")
+	if i < 0 {
+		return nil, fmt.Errorf("registry URI %q could not be parsed, make sure it is the fully qualified registry URI without the scheme/protocol. e.g domain.com:80/my/repository:digest", registryUri)
+	}
+
+	registryScope := registryUri[:i]
+	if err := validateRegistryScopeFormat(registryScope); err != nil {
+		return nil, err
+	}
```
Comment (gokarnm) [inline verification/policy.go:219]:
```suggestion
func (policyDoc *PolicyDocument) getApplicableTrustPolicy(artifactUri string) (*TrustPolicy, error) {
	i := strings.LastIndex(artifactUri, ":")
	if i < 0 {
		return nil, fmt.Errorf("artifactUri URI %q could not be parsed, make sure it is the fully qualified OCI registry artifact URI without the scheme/protocol. e.g domain.com:80/my/repository:digest", artifactUri)
	}

	artifactPath := registryUri[:i]
	if err := validateRegistryScopeFormat(artifactPath); err != nil {
		return nil, err
	}
```

File: verification/policy.go [hunk 4]
```diff
@@ -203,3 +203,44 @@ func ValidatePolicyDocument(policyDoc *PolicyDocument) error {
...
+		}
+	}
+
+	if applicablePolicy != nil {
+		// a policy with exact match for registry URI takes precedence over a wildcard (*) policy.
+		return applicablePolicy, nil
+	} else if wildcardPolicy != nil {
+		return wildcardPolicy, nil
+	} else {
+		return nil, fmt.Errorf("registry scope %q has no applicable trust policy", artifactPath)
```
Comment (gokarnm) [inline verification/policy.go:233]:
```suggestion
		return nil, fmt.Errorf("artifact %q has no applicable trust policy", artifactUri)
```

File: verification/helpers.go [hunk 5]
```diff
@@ -18,6 +18,20 @@ func isPresent(val string, values []string) bool {
 	return false
 }
 
+func getArtifactPathFromUri(artifactUri string) (string, error) {
+	// TODO support more types of URI like "domain.com/repository", "domain.com/repository:tag", "domain.com/repository@sha256:digest"
+	i := strings.LastIndex(artifactUri, ":")
+	if i < 0 {
+		return "", fmt.Errorf("registry URI %q could not be parsed, make sure it is the fully qualified registry URI without the scheme/protocol. e.g domain.com:80/my/repository:digest", artifactUri)
```
Comment (gokarnm) [inline verification/helpers.go:25]:
```suggestion
		return "", fmt.Errorf("artifact URI %q could not be parsed, make sure it is the fully qualified OCI artifact URI without the scheme/protocol. e.g domain.com:80/my/repository:digest", artifactUri)
```

**Review by gokarnm — CHANGES_REQUESTED** [review]

File: verification/policy.go [hunk 6]
```diff
@@ -203,3 +203,44 @@ func ValidatePolicyDocument(policyDoc *PolicyDocument) error {
...
+		}
+	}
+
+	if applicablePolicy != nil {
+		// a policy with exact match for registry URI takes precedence over a wildcard (*) policy.
+		return applicablePolicy, nil
+	} else if wildcardPolicy != nil {
+		return wildcardPolicy, nil
+	} else {
+		return nil, fmt.Errorf("artifact %q has no applicable trust policy", artifactPath)
```
Comment (gokarnm) [inline verification/policy.go:233]: Nit, I'd prefer `artifactUri` in the error message which is the user provided input, `artifactPath` is calculated in our internal logic.

```suggestion
		return nil, fmt.Errorf("artifact %q has no applicable trust policy", artifactUri)
```

**Review by gokarnm — APPROVED** [review]

File: verification/policy.go [hunk 7]
```diff
@@ -135,10 +135,10 @@ func validateTrustStore(statement TrustPolicy) error {
 
 // ValidatePolicyDocument validates a policy document according to it's version's rule set.
 // if any rule is violated, returns an error
-func ValidatePolicyDocument(policyDoc *PolicyDocument) error {
+func (policyDoc *PolicyDocument) ValidatePolicyDocument() error {
 	// Constants
 	supportedPolicyVersions := []string{"1.0"}
-	supportedVerificationPresets := []string{"strict", "permissive", "audit", "skip"}
+	supportedVerificationLevels := []string{"strict", "permissive", "audit", "skip"}
```
Comment (priteshbandi) [inline verification/policy.go:141]: Are you going to replaces this with levels define in https://github.com/notaryproject/notation-go/pull/55 ?

File: verification/policy.go [hunk 8]
```diff
@@ -203,3 +203,44 @@ func ValidatePolicyDocument(policyDoc *PolicyDocument) error {
...
+	}
+
+	var wildcardPolicy *TrustPolicy
+	var applicablePolicy *TrustPolicy
+	for _, policyStatement := range policyDoc.TrustPolicies {
+		if isPresent(wildcard, policyStatement.RegistryScopes) {
+			wildcardPolicy = policyStatement.deepCopy() // we need to deep copy because we can't use the loop variable address. see https://stackoverflow.com/a/45967429
+		} else if isPresent(artifactPath, policyStatement.RegistryScopes) {
+			applicablePolicy = policyStatement.deepCopy()
+		}
```
Comment (priteshbandi) [inline verification/policy.go:224]: qq: Where are we validating that there is only one applicable trust policy fir a given repository?

File: verification/policy.go (hunk 7 above)
Comment (rgnote) [reply verification/policy.go:141]: Yes. I have a TODO to do that.

File: verification/policy.go (hunk 8 above)
Comment (rgnote) [reply verification/policy.go:224]: https://github.com/notaryproject/notation-go/blob/main/verification/policy.go#L69

**Review by priteshbandi — APPROVED** [review]
LGTM

**Comment by gokarnm** [comment]
@shizhMSFT merging this PR with two approvals.

**Pull request merged**